File: 17-07-2025_Find_the_Maximum_Length_of_Valid_Subsequence_II.cpp
class Solution {
public:
    
    int maximumLength(vector<int>& nums, int k) {
        int ans = 1;
        for (int val = 0; val < k; ++val) {
            vector<int> dp(k, 0);
        for (int num : nums) {
            int rem = num % k;
            int required_rem = (k + val - rem) % k;
            dp[rem] = max(dp[rem], dp[required_rem] + 1);
            ans = max(ans, dp[rem]);
        }
    }

    return ans;
}
};

File: 17-07-2025_Find_the_Maximum_Length_of_Valid_Subsequence_II.py
class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        ans = 1

    # Try all possible values of (sub[i] + sub[i+1]) % k
        for val in range(k):
            dp = [0] * k  # dp[i] = max length of subsequence ending with num % k == i

            for num in nums:
                rem = num % k
                required_rem = (k + val - rem) % k
                dp[rem] = max(dp[rem], dp[required_rem] + 1)
                ans = max(ans, dp[rem])

        return ans        
